rfd/0228-resource-scoped-constraints-in-access-requests.md

```diff
@@ -0,0 +1,198 @@
+
+#### Validation & Resolution
+Resource-based Access Requests including constraints are validated using constraint-aware role matchers derived from each `ResourceID.Constraints` (e.g., ARN/permission set/SSH login matchers). `RequestValidator` extends the existing `applicableSearchAsRoles`/`roleAllowsResource` flow by passing these matchers, so a role only qualifies if it allows the requested resource *and* the specified constraints. Resources are still pre-filtered by current access and requested `ResourceIDs` are still deduplicated and size-capped, as they are currently. The resulting role list is the set of requestable/search-as roles that satisfies all `(resource, constraints)` pairs (when multiple roles qualify, behavior matches current semantics). If no roles qualify, existing failure behavior and errors apply.
+
+#### Authorization
+On approval/assumption, certs carry the exact `(resource, constraints)` pairs via `tlsca.Identity.AllowedResourceIDs` (each entry including `ResourceID.Constraints`). `AccessChecker` enforces those constraints: if a role grants multiple principals on a resource, approving a request for a narrower set (e.g., a single ARN, specific SSH logins, or IC permission sets) must only allow those requested items. By reading `ResourceID.Constraints` from `AllowedResourceIDs`, `AccessChecker` scopes evaluation accordingly and denies any not-requested principals even if the resolved role(s) in the cert would otherwise permit them.
+
+### Web/API (high-level)
+- Listing `UnifiedResources` surfaces per-constraint status when "show requestable" is enabled. For each item (ARN, permission set, SSH login), a `requiresRequest: true|false` marker is provided so the UI can render a single menu with "Granted" and "Requestable" sections.
+- Resource Access Requests that include `ResourceID.Constraints` present both requested constraints and resolved roles to reviewers.
+
+## Compatibility
+- The `constraints` field is optional; existing clients and requests without constraints continue to work unchanged.
+- UI can adopt the per-constraint "granted vs requestable" grouping incrementally by resource type.
+- RBAC semantics are additive. Constraint matchers are considered only when constraints are present in a request.
+- Serialized `ResourceID`s with constraints will not be parsable by older versions. This is an intentional fail-closed design to avoid potential over-granting if constraints are ignored.
+  - For now, to keep non-constraint-bearing `ResourceID`s compatible with older agents, we retain the existing slash-delimited string format for those, and only use the JSON format when constraints are present. The next major version could consider switching fully to JSON for all `ResourceID`s to simplify code paths.
```

Comment:
I'm still not entirely sure what this means exactly, i.e. how we're going to ensure backwards compatibility. What does it mean "retain slash-delimited string format for older agents"? Let's work through a specific example:

- I have a cluster and agents on older version (prior to this change).
- I upgrade control plane (auth, proxy) to the new version with this change.
- I now need access to SSH node-1 as "root". It may be an old agent or a new one.

What happens in this case, what does the certificate look like after an access request is created and approved?